client/src/dom-tagging.js

```diff
@@ -0,0 +1,344 @@
+    console.log('DOMT debug', 'hijack add event listener', 'this:', this, 'eventName:', eventName, 'eventHandler:', eventHandler)
+    initialiseEventLog(eventName)
+
+    oldAddEventListener.call(
+      this, eventName, (event) => {
+        logEvent(eventName, event)
+        eventHandler(event)
+      }, false // default
+    )
+  }
+}
+*/
+
+const getEventTrace = () => {
+  const l = eventLog
+  if (!l.domModifyingEvents) return null
+  const events = l.domModifyingEvents.map(
+    (dme) => {
+      const historyObject = l.eventHistory[dme.historyIndex]
+      // Get what key to find correct event type array
+      const eventLogKey = historyObject.eventName
+      // Get index in that array for correct event object
+      const eventLogArrayIndex = historyObject.index
+      const event = l[eventLogKey][eventLogArrayIndex]
+      let { path } = event
+      if (!path) path = eventPathFallback(event)
+      console.log('getTrace', 'event.path:', event.path, 'eventPathFallback', eventPathFallback(event))
+      return getXPath(path[0])
+    }
+  )
+  return events
+}
+
+// Tagging
+
+/*const localStorageKey = 'swp1-tagging-concept'
+
+const getCommentsArray = () => JSON.parse(localStorage.getItem(localStorageKey)) || []
+
+const attributeName = 'data-comment-count'
+
+// Loading comments
+
+const simulateEvents = (trace) => {
+  if (trace) {
+    console.log('📲', 'Simulate events from trace', trace)
+    trace.forEach(
+      (step) => {
+        getElementByXPath(step).click()
+      }
+    )
+  }
+}
+
+const loadTags = () => {
+  /*
+    const items = getCommentsArray()
+    console.log('💾', 'getCommentsArray in loadTags =>  items', items)
+    items.forEach((item) => {
+      let el = getElementByXPath(item.path)
+      if (!el) {
+        // Element is not in DOM
+        // Simulate events
+        simulateEvents(item.trace)
+        el = getElementByXPath(item.path)
+      }
+      if (el) {
+        let value = el.getAttribute(attributeName) || 0
+        el.setAttribute(attributeName, ++value)
+        let title = el.getAttribute('title') || ''
+        title = title.length ? `${title}, ` : title
+        el.setAttribute('title', title + item.comment)
+      }
+    })
+}
+
+const refreshTags = () => {
+
+    ['title', attributeName].forEach((a) => {
+      document.querySelectorAll(`[${a}]`).forEach(el => el.removeAttribute(a))
+    })
+    loadTags()
+
+}
+
+// Save comment:
+
+const saveTag = (el) => {
+
+    const path = getXPath(el)
+    const comment = prompt('Add comment')
+    const trace = getEventTrace()
+    console.log('🛤️', 'Tagged element path', path, 'trace for events', trace)
+    const items = getCommentsArray()
+    // Add to array
+    items.push({
+      path,
+      trace,
+      comment,
+    })
+    // Store
+    localStorage.setItem(
+      localStorageKey,
+      JSON.stringify(items)
+    )
+    // Refresh view
+    refreshTags()
+
+}
+*/
```

Comment:
Same as above, dead code or will be implemented in the future?

Reply:
will be implemented